sdcm/remote.py

```diff
@@ -765,6 +725,237 @@ def _make_rsync_cmd(self, src, dst, delete_dst, preserve_symlinks, timeout=300):
+        self._next_cleanup = time.time() + self._connection_thread_map_cleanup_frequency
+        super(RemoteLibSSH2CmdRunner, self)._prepare()
+
+    def _create_connection(self):
+        self._cleanup_connection_thread_map()
+        return LibSSH2Client(
+            host=self.hostname,
+            user=self.user,
+            port=self.port,
+            pkey=os.path.expanduser(self.key_file),
+            timings=Timings(keepalive_timeout=300, connect_timeout=self.connect_timeout)
+        )
+
+    @retrying(n=3, sleep_time=5, allowed_exceptions=(RetriableNetworkException,))
+    def run(self, cmd, timeout=None, ignore_status=False,  # pylint: disable=too-many-arguments
+            verbose=True, new_session=False, log_file=None, retry=1, watchers=None):
+
+        watchers = self._setup_watchers(verbose=verbose, log_file=log_file, additional_watchers=watchers)
+
+        @retrying(n=retry)
+        def _run():
+            try:
+                if verbose:
+                    self._log.debug('Running command "{}"...'.format(cmd))
+                start_time = time.time()
+                command_kwargs = dict(
+                    command=cmd, warn=ignore_status,
+                    encoding='utf-8', hide=True,
+                    watchers=watchers, timeout=timeout,
+                    in_stream=False
+                )
+                if new_session:
+                    with self._create_connection() as connection:
+                        result = connection.run(**command_kwargs)
+                else:
+                    result = self._connection.run(**command_kwargs)
+                setattr(result, 'duration', time.time() - start_time)
+                setattr(result, 'exit_status', result.exited)
+                return result
+            except NETWORK_EXCEPTIONS as ex:
+                LOGGER.error(ex)
+                if isinstance(ex, FailedToRunCommand) and not new_session:
+                    self._log.debug('Reestablish the session...')
+                    try:
+                        self._connection.disconnect()
+                    except:  # pylint: disable=bare-except
+                        pass
+                    try:
+                        self._connection.connect()
+                    except:  # pylint: disable=bare-except
+                        pass
+                if is_error_retryable(str(ex)) or isinstance(ex, LIBSSH2_RETRYABLE_EXCEPTIONS):
+                    raise RetriableNetworkException(str(ex), original=ex)
+                raise
+            except Exception as details:  # pylint: disable=broad-except
+                if hasattr(details, "result"):
+                    self._print_command_results(details.result, verbose, ignore_status)  # pylint: disable=no-member
+                raise
+
+        result = _run()
+        self._print_command_results(result, verbose, ignore_status)
+        return result
+
+    def is_up(self, timeout=30):
+        end_time = time.time() + timeout
```

Comment:
Why ? Only difference i know is that it has undefined static starting point. Probably monotic would be better idea

Reply:
In few words `perf_counter()` is the same as `monotonic()` but with highest available precision and there is no performance penalty for using `perf_counter()` instead of `monotonic()`

Good explanation here: https://www.webucator.com/blog/2015/08/python-clocks-explained/

Reply:
If it is monotic, it make sense, addressing that